File: src/main/java/dev/tensorhero/tinytorch/Losses.java
package dev.tensorhero.tinytorch;

import dev.tensorhero.tinynum.NDArray;

/**
 * Losses — loss functions for training neural networks.
 *
 * <p>All methods are static and return a scalar {@link Tensor} of shape [1].
 * Forward-only in E04; backward implementations will be added in E06.</p>
 */
public class Losses {

    // ================================================================
    // E04 — Loss Functions
    // ================================================================

    /**
     * Numerically stable log-softmax along the given axis.
     *
     * <p>logSoftmax(x_i) = x_i - max(x) - log(sum(exp(x - max(x))))</p>
     */
    public static Tensor logSoftmax(Tensor x, int axis) {
        throw new UnsupportedOperationException("TODO: E04");
    }

    /**
     * Mean Squared Error: mean((pred - target)^2).
     *
     * @return scalar Tensor of shape [1]
     */
    public static Tensor mse(Tensor pred, Tensor target) {
        throw new UnsupportedOperationException("TODO: E04");
    }

    /**
     * Cross-entropy loss with one-hot encoded targets.
     *
     * <p>CE = -mean(sum(targets * logSoftmax(logits), axis=classes))</p>
     *
     * @param logits  raw scores, shape [batch, numClasses]
     * @param targets one-hot encoded, shape [batch, numClasses]
     * @return scalar Tensor of shape [1]
     */
    public static Tensor crossEntropy(Tensor logits, Tensor targets) {
        throw new UnsupportedOperationException("TODO: E04");
    }

    // ================================================================
    // E06 — More Backward Ops
    // ================================================================
    // crossEntropy: when logits.requiresGrad, create CrossEntropyBackward,
    //   call fn.forward(), set result.requiresGrad = true, result.gradFn = fn.
    //   Uses fused gradient: dL/dlogits = (softmax(logits) - targets) / batchSize.
    // mse: rewrite to use Tensor ops (sub → mul → reshape → mean) so the
    //   computation graph is recorded automatically. No MSEBackward needed.
}

File: tinytorch/losses.py
"""Losses — loss functions for training neural networks.

All functions return a scalar Tensor of shape [1].
Forward-only in E04; backward implementations will be added in E06.
"""

from __future__ import annotations

from tinynum import NDArray
from tinytorch.tensor import Tensor


def log_softmax(x: Tensor, axis: int) -> Tensor:
    """Numerically stable log-softmax along the given axis.

    logSoftmax(x_i) = x_i - max(x) - log(sum(exp(x - max(x))))
    """
    raise NotImplementedError("TODO: E04")


def mse(pred: Tensor, target: Tensor) -> Tensor:
    """Mean Squared Error: mean((pred - target)^2).

    Returns:
        Scalar Tensor of shape [1].
    """
    raise NotImplementedError("TODO: E04")


def cross_entropy(logits: Tensor, targets: Tensor) -> Tensor:
    """Cross-entropy loss with one-hot encoded targets.

    CE = -mean(sum(targets * logSoftmax(logits), axis=classes))

    Args:
        logits: raw scores, shape [batch, num_classes].
        targets: one-hot encoded, shape [batch, num_classes].

    Returns:
        Scalar Tensor of shape [1].
    """
    raise NotImplementedError("TODO: E04")


# ================================================================
# E06 — More Backward Ops
# ================================================================
# cross_entropy: when logits.requires_grad, create CrossEntropyBackward,
#   call fn.forward(), set result.requires_grad = True, result.grad_fn = fn.
#   Uses fused gradient: dL/dlogits = (softmax(logits) - targets) / batch_size.
# mse: rewrite to use Tensor ops (sub → mul → reshape → mean) so the
#   computation graph is recorded automatically. No MSEBackward needed.
